packages/Search/src/details/DTK_DetailsTmp.hpp

```diff
@@ -0,0 +1,690 @@
+    ////////////////////////////////////////////////////////////////////////////
+    tree._upper_tree->query( queries, indices, offset );
+    ////////////////////////////////////////////////////////////////////////////
+
+    ////////////////////////////////////////////////////////////////////////////
+    // Forward queries
+    ////////////////////////////////////////////////////////////////////////////
+    Kokkos::View<int *, DeviceType> ids( "ids" );
+    Kokkos::View<Query *, DeviceType> fwd_queries( "receive_buffer" );
+    forward_queries( tree._comm, queries, indices, offset, fwd_queries, ids,
+                     ranks );
+    ////////////////////////////////////////////////////////////////////////////
+
+    ////////////////////////////////////////////////////////////////////////////
+    // Perform queries that have been received
+    ////////////////////////////////////////////////////////////////////////////
+    tree._bvh.query( fwd_queries, indices, offset );
+    ////////////////////////////////////////////////////////////////////////////
+
+    ////////////////////////////////////////////////////////////////////////////
+    // Communicate results back
+    ////////////////////////////////////////////////////////////////////////////
+    communicate_results_back( tree._comm, indices, offset, ranks, ids );
+    ////////////////////////////////////////////////////////////////////////////
+
+    ////////////////////////////////////////////////////////////////////////////
+    // Merge results
+    ////////////////////////////////////////////////////////////////////////////
+    int const n_queries = queries.extent_int( 0 );
+    count_results( n_queries, ids, offset );
+    sort_results( ids, indices, ranks );
+    ////////////////////////////////////////////////////////////////////////////
+}
+
+template <typename DeviceType>
+void Tmp<DeviceType>::sort_results(
+    Kokkos::View<int *, DeviceType> query_ids,
+    Kokkos::View<int *, DeviceType> results,
+    Kokkos::View<int *, DeviceType> ranks,
+    Kokkos::View<double *, DeviceType> distances )
+{
+    int const n = query_ids.extent( 0 );
+
+    typedef Kokkos::BinOp1D<Kokkos::View<int *, DeviceType>> CompType;
+
+    Kokkos::Experimental::MinMaxScalar<int> result;
+    Kokkos::Experimental::MinMax<int> reducer( result );
+    parallel_reduce(
+        Kokkos::RangePolicy<ExecutionSpace>( 0, n ),
+        Kokkos::Impl::min_max_functor<Kokkos::View<int *, DeviceType>>(
+            query_ids ),
+        reducer );
+    if ( result.min_val == result.max_val )
+        return;
+    Kokkos::BinSort<Kokkos::View<int *, DeviceType>, CompType> bin_sort(
+        query_ids, CompType( n / 2, result.min_val, result.max_val ), true );
+    bin_sort.create_permute_vector();
+    bin_sort.sort( results );
+    bin_sort.sort( ranks );
+    bin_sort.sort( distances );
+}
+
+template <typename DeviceType>
+void Tmp<DeviceType>::sort_results( Kokkos::View<int *, DeviceType> query_ids,
```

Comment:
I know.  Same thing with `communicate_results_back()`.  One option might be creating some kind of struct encapsulating the sort with respect of ids and that has a sort member function templated on the `T` in `View<T*, ...>`.